### PR TITLE
[SPARK-40054][SQL] Restore the error handling syntax of try_cast()

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

For the following query
```
SET spark.sql.ansi.enabled=true;
SELECT try_cast(1/0 AS string); 
```
Spark 3.3 will throw an exception for the division by zero error. In the current master branch, it returns null after the refactoring PR https://github.com/apache/spark/pull/36703

This PR is to restore the previous error handling syntax.

### Why are the changes needed?

1. Restore the behavior of try_cast()
2. The previous syntax is more reasonable. Users can cleanly catch the exception from the child of `try_cast`.

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by gengliangwang** [comment]
Some of the code are from the removed part of https://github.com/apache/spark/pull/36703, which is still reliable.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TryCastSuite.scala [hunk 1]
```diff
@@ -17,30 +17,50 @@
...
+
 import org.apache.spark.SparkFunSuite
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.util.DateTimeTestUtils.UTC_OPT
 import org.apache.spark.sql.types._
 import org.apache.spark.unsafe.types.UTF8String
 
 // A test suite to check analysis behaviors of `TryCast`.
-class TryCastSuite extends SparkFunSuite {
+class TryCastSuite extends CastWithAnsiOnSuite {
```
Comment (gengliangwang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TryCastSuite.scala:29]: This recovers the test coverage before https://github.com/apache/spark/pull/36703

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 2]
```diff
@@ -460,29 +500,57 @@ case class Cast(
...
     } else {
       TypeCheckResult.TypeCheckFailure(typeCheckFailureMessage)
     }
   }
 
-  override def nullable: Boolean = child.nullable || Cast.forceNullable(child.dataType, dataType)
+  override def nullable: Boolean = if (!isTryCast) {
+    child.nullable || Cast.forceNullable(child.dataType, dataType)
+  } else {
+    (child.dataType, dataType) match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:545]: is this new?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 3]
```diff
@@ -1247,11 +1329,22 @@ case class Cast(
   protected[this] def castCode(ctx: CodegenContext, input: ExprValue, inputIsNull: ExprValue,
     result: ExprValue, resultIsNull: ExprValue, resultType: DataType, cast: CastFunction): Block = {
     val javaType = JavaCode.javaType(resultType)
+    val addTryCatchOnCastIfNeeded = if (!isTryCast) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:1332]: this is more like a method name. how about `castCodeWithTryCatchIfNeeded`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 4]
```diff
@@ -1795,9 +1795,9 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
         cast
 
       case SqlBaseParser.TRY_CAST =>
-        // `TryCast` can only be user-specified and we don't need to set the USER_SPECIFIED_CAST
-        // tag, which is only used by `Cast`
-        TryCast(expression(ctx.expression), dataType)
+        val cast = Cast(expression(ctx.expression), dataType, evalMode = EvalMode.TRY)
+        cast.setTagValue(Cast.USER_SPECIFIED_CAST, true)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1799]: do we need to set this? Cast with `TRY` mode can only be user specified

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TryCastSuite.scala [hunk 5]
```diff
@@ -79,3 +99,14 @@ class TryCastSuite extends SparkFunSuite {
     assert(!c4.resolved)
   }
 }
+
+class TryCastThrowExceptionSuite extends SparkFunSuite with ExpressionEvalHelper {
+  // The method checkExceptionInExpression is overridedn in TryCastSuite, so here we have a
+  // new test suite for testing exceptions from the child of `try_cast()`.
+  test("TryCast should not catch the exception from it's child") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TryCastSuite.scala:106]: why can't we put this test in `TryCastSuite`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:545]: Yes

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 4 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1799]: try_cast is using exactly the expression `CAST` now. Better set the tag as well.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TryCastSuite.scala (hunk 5 above)
Comment (gengliangwang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TryCastSuite.scala:106]: Please read my comment above. 
The method checkExceptionInExpression is overridden in TryCastSuite for testing all the cases should return nulls.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 4 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1799]: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala#L122 here it only matches CAST

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 6]
```diff
@@ -460,29 +500,57 @@ case class Cast(
...
-  override def nullable: Boolean = child.nullable || Cast.forceNullable(child.dataType, dataType)
+  override def nullable: Boolean = if (!isTryCast) {
+    child.nullable || Cast.forceNullable(child.dataType, dataType)
+  } else {
+    (child.dataType, dataType) match {
+      case (StringType, BinaryType) => child.nullable
+      // TODO: Implement a more accurate method for checking whether a decimal value can be cast
+      //       as integral types without overflow. Currently, the cast can overflow even if
+      //       "Cast.canUpCast" method returns true.
+      case (_: DecimalType, _: IntegralType) => true
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:550]: is this a bug in spark 3.3?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TryCastSuite.scala [hunk 7]
```diff
@@ -17,30 +17,50 @@
...
+  override def cast(v: Any, targetType: DataType, timeZoneId: Option[String] = None): Cast = {
     v match {
-      case lit: Expression => TryCast(lit, targetType, timeZoneId)
-      case _ => TryCast(Literal(v), targetType, timeZoneId)
+      case lit: Expression => Cast(lit, targetType, timeZoneId, EvalMode.TRY)
+      case _ => Cast(Literal(v), targetType, timeZoneId, EvalMode.TRY)
     }
   }
 
+  override def checkExceptionInExpression[T <: Throwable : ClassTag](
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TryCastSuite.scala:40]: this is a bit weird, I see code like this
```
    if (!isTryCast) {
      checkExceptionInExpression[ArithmeticException](ret, "overflow")
    } else {
      checkEvaluation(ret, Row(null))
    }
```

if we override `checkExceptionInExpression`, why do we still need `if-else`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 6 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:550]: Decimal(10, 0) is considered as "canUpCast" as Int. However, `Int.MaxValue + 1` can fit in Decimal(10, 0) too..
It has been a long time and the strict table insertion policy relies on this. 
I would consider fixing it in another PR. For this one I will leave it as it is now.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TryCastSuite.scala (hunk 7 above)
Comment (gengliangwang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TryCastSuite.scala:40]: That one was "Row(null)" with schema `struct<a:int>`. Different from the default answer `null`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 6 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:550]: Note that this is not a bug in Spark 3.3. I am avoid making the nullability of try_cast as always null here.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TryCastSuite.scala (hunk 7 above)
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/TryCastSuite.scala:40]: It's a bit weird to have two different testing approaches for try_cast, but given that it was already the case before, I don't have a strong opinion.

**Review by cloud-fan — APPROVED** [review]

**Comment by gengliangwang** [comment]
@cloud-fan Thanks for the review. Merging to master.

**Pull request closed without merging**